core/common/src/main/java/alluxio/underfs/UnderFileSystemFactoryRegistry.java

```diff
@@ -44,58 +28,20 @@
- * standard implementations, depending on the order in which the JVM discovers the services your own
- * implementation may not take priority. You can enable {@code DEBUG} level logging for this class
- * to see the order in which factories are discovered and which is selected when obtaining a
- * {@link UnderFileSystem} instance for a path. If this shows that your implementation is not
- * getting discovered or used, you may wish to use the manual registration approach.
- * </p>
- * <h4>Automatic Discovery</h4>
- * <p>
- * To use the {@linkplain ServiceLoader} based mechanism you need to have a file named
- * {@code alluxio.underfs.UnderFileSystemFactory} placed in the {@code META-INF\services} directory
- * of your project. This file should contain the full name of your factory types (one per line),
- * your factory types must have a public unparameterised constructor available (see
- * {@link ServiceLoader} for more detail on this). You can enable {@code DEBUG} level logging to see
- * factories as they are discovered if you wish to check that your implementation gets discovered.
- * </p>
  * <p>
  * Note that if you are bundling Alluxio plus your code in a shaded JAR using Maven, make sure to
  * use the {@code ServicesResourceTransformer} as otherwise your services file will override the
  * core provided services file and leave the standard factories and under file system
  * implementations unavailable.
  * </p>
- * <h4>Manual Registration</h4>
- * <p>
- * To manually register a factory, simply pass an instance of your factory to the
- * {@link #register(UnderFileSystemFactory)} method. This can be useful when your factory cannot be
- * instantiated without arguments or in cases where automatic discovery does not give your factory
- * priority. Factories registered this way will be registered at the start of the factories list so
- * will have the first opportunity to indicate whether they support a requested path.
- * </p>
+ * @see ExtensionFactoryRegistry
  */
 @NotThreadSafe
 public final class UnderFileSystemFactoryRegistry {
   private static final Logger LOG = LoggerFactory.getLogger(UnderFileSystemFactoryRegistry.class);
-
-  /**
-   * The base list of factories, which does not include any lib or extension factories. The only
-   * factories in the base list will be the LocalUFS factory, and any additional factories
-   * registered by tests. All other UFS factories will be discovered and service loaded when ufs
-   * creation occurs.
-   */
-  private static final List<UnderFileSystemFactory> FACTORIES = new CopyOnWriteArrayList<>();
-
-  public static final String LIB_DIR =
-      PathUtils.concatPath(Configuration.get(PropertyKey.HOME), "lib");
-
-  private static boolean sInit = false;
+  private static final String EXTENSION_PATTERN = "alluxio-underfs-*.jar";
+  private static ExtensionFactoryRegistry<UnderFileSystemFactory, UnderFileSystemConfiguration>
```

Comment:
how are you supposed to use `ExtensionFactoryRegistry`? extend it or create one, or create a static one?

Reply:
Normally you can just create one and hold on to it. For `UnderFileSystemFactory` there are many integration tests reply on the registry being static to register test factories, so we have to keep it static.